File: backend/pkg/canonical/hmac.go
package canonical

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"errors"
)

var (
	// ErrInvalidSignature indicates signature verification failed
	ErrInvalidSignature = errors.New("invalid HMAC signature")
)

// SignHMAC signs the signature subset using HMAC-SHA256.
//
// Process (WP2 simplified):
//  1. Generate canonical signature payload
//  2. HMAC-SHA256 the payload directly with provided key
//  3. Return base64-encoded signature
//
// Pre-hashing is unnecessary since HMAC provides cryptographic security.
//
// Args:
//
//	subset: SignatureSubset with required fields
//	key: HMAC secret key (bytes)
//
// Returns:
//
//	Base64-encoded HMAC signature, or error if payload generation fails
//
// Example:
//
//	subset := &SignatureSubset{
//	    PCSID:      "test",
//	    DHat:       1.0,
//	    CohStar:    0.75,
//	    R:          0.5,
//	    Budget:     0.35,
//	    MerkleRoot: "abc",
//	    Epoch:      1,
//	    ShardID:    "s1",
//	}
//	signature, err := SignHMAC(subset, []byte("my-secret-key"))
func SignHMAC(subset *SignatureSubset, key []byte) (string, error) {
	payload, err := CanonicalJSONBytes(subset)
	if err != nil {
		return "", err
	}

	mac := hmac.New(sha256.New, key)
	mac.Write(payload)
	sig := mac.Sum(nil)

	return base64.StdEncoding.EncodeToString(sig), nil
}

// VerifyHMAC verifies an HMAC-SHA256 signature.
//
// Process (WP2 simplified):
//  1. Generate canonical signature payload
//  2. HMAC-SHA256 the payload directly with provided key
//  3. Decode base64 signature and compare using constant-time comparison
//
// Args:
//
//	subset: SignatureSubset with required fields
//	sigB64: Base64-encoded HMAC signature
//	key: HMAC secret key (bytes)
//
// Returns:
//
//	nil if verification succeeds, error otherwise
func VerifyHMAC(subset *SignatureSubset, sigB64 string, key []byte) error {
	payload, err := CanonicalJSONBytes(subset)
	if err != nil {
		return err
	}

	// Compute expected HMAC directly on payload
	mac := hmac.New(sha256.New, key)
	mac.Write(payload)
	expected := mac.Sum(nil)

	// Decode provided signature
	got, err := base64.StdEncoding.DecodeString(sigB64)
	if err != nil {
		return err
	}

	// Constant-time comparison
	if !hmac.Equal(expected, got) {
		return ErrInvalidSignature
	}

	return nil
}

File: sdk/ts/src/hmac.ts
/**
 * HMAC-SHA256 signing utilities for PCS.
 * Phase 10 WP2: Simplified HMAC (sign payload directly, no pre-hash).
 */

import { createHmac } from 'crypto';
import { signaturePayload } from './canonical.js';

/**
 * Sign a PCS using HMAC-SHA256.
 *
 * Process (WP2 simplified):
 * 1. Generate canonical signature payload
 * 2. HMAC-SHA256 the payload directly with provided key
 * 3. Return base64-encoded signature
 *
 * Pre-hashing is unnecessary since HMAC provides cryptographic security.
 *
 * @param pcs - Complete PCS object
 * @param key - HMAC secret key (string or Buffer)
 * @returns Base64-encoded HMAC signature
 *
 * @example
 * ```typescript
 * const pcs = {
 *   pcs_id: "test",
 *   D_hat: 1.0,
 *   coh_star: 0.75,
 *   r: 0.5,
 *   budget: 0.35,
 *   merkle_root: "abc",
 *   epoch: 1,
 *   shard_id: "s1"
 * };
 * const signature = signHMAC(pcs, "my-secret-key");
 * ```
 */
export function signHMAC(pcs: Record<string, any>, key: string | Buffer): string {
  const payload = signaturePayload(pcs);
  const hmac = createHmac('sha256', key);
  hmac.update(payload);
  return hmac.digest('base64');
}

/**
 * Verify HMAC-SHA256 signature on a PCS.
 *
 * Process (WP2 simplified):
 * 1. Extract signature from pcs.sig
 * 2. Generate canonical signature payload (excluding sig field)
 * 3. HMAC-SHA256 the payload directly with provided key
 * 4. Compare with extracted signature
 *
 * @param pcs - Complete PCS object with "sig" field
 * @param key - HMAC secret key (string or Buffer)
 * @returns true if signature is valid, false otherwise
 *
 * @example
 * ```typescript
 * const pcs = {
 *   pcs_id: "test",
 *   D_hat: 1.0,
 *   coh_star: 0.75,
 *   r: 0.5,
 *   budget: 0.35,
 *   merkle_root: "abc",
 *   epoch: 1,
 *   shard_id: "s1",
 *   sig: "base64-signature"
 * };
 * const isValid = verifyHMAC(pcs, "my-secret-key");
 * ```
 */
export function verifyHMAC(pcs: Record<string, any>, key: string | Buffer): boolean {
  try {
    // Extract signature
    const sigB64 = pcs.sig;
    if (!sigB64) {
      return false;
    }

    // Compute signature over PCS (excluding sig field)
    const pcsWithoutSig = { ...pcs };
    delete pcsWithoutSig.sig;

    const payload = signaturePayload(pcsWithoutSig);
    const hmac = createHmac('sha256', key);
    hmac.update(payload);
    const computed = hmac.digest('base64');

    // Simple string comparison (Node.js crypto.timingSafeEqual would be better)
    return computed === sigB64;
  } catch (error) {
    // Any error in verification should return false
    return false;
  }
}
